execution/executor/src/components/in_memory_state_calculator.rs

```diff
@@ -25,25 +25,43 @@ use std::{
     sync::Arc,
 };
 use storage_interface::{
-    in_memory_state::InMemoryState, verified_state_view::StateCache, DbReader, TreeState,
+    in_memory_state::InMemoryState,
+    verified_state_view::{StateCache, VerifiedStateView},
+    DbReader, TreeState,
 };
 
 pub trait IntoLedgerView {
     fn into_ledger_view(self, db: &Arc<dyn DbReader>) -> Result<ExecutedTrees>;
 }
 
 impl IntoLedgerView for TreeState {
-    fn into_ledger_view(self, _db: &Arc<dyn DbReader>) -> Result<ExecutedTrees> {
-        let checkpoint_num_txns = self.num_transactions;
+    fn into_ledger_view(self, db: &Arc<dyn DbReader>) -> Result<ExecutedTrees> {
+        let checkpoint_num_txns = self
+            .num_transactions
+            .checked_sub(self.write_sets_after_checkpoint.len() as LeafCount)
+            .ok_or_else(|| anyhow!("State checkpoint pre-dates genesis."))?;
         let checkpoint =
```

Comment:
nit: `in_memory_state`  instead of `checkpoint` as you have a `checkpoint` inside this and this is not the checkpoint.